File: test/benchmark/kostya/collatz.js
// Kostya Benchmark: collatz (Node.js)
// Longest Collatz (3n+1) sequence under 1,000,000
'use strict';

function collatzLen(n) {
    let steps = 1;
    let x = n;
    while (x !== 1) {
        if (x % 2 === 0) {
            x = x / 2;  // avoid >> which truncates to 32-bit
        } else {
            x = 3 * x + 1;
        }
        steps++;
    }
    return steps;
}

function benchmark() {
    const limit = 1000000;
    let maxLen = 0;
    let maxStart = 0;
    for (let i = 1; i < limit; i++) {
        const clen = collatzLen(i);
        if (clen > maxLen) {
            maxLen = clen;
            maxStart = i;
        }
    }
    return maxStart;
}

function main() {
    const result = benchmark();
    if (result === 837799) {
        process.stdout.write("collatz: PASS (start=" + result + ")\n");
    } else {
        process.stdout.write("collatz: FAIL result=" + result + "\n");
    }
}

main();

File: test/benchmark/kostya/python/collatz.py
#!/usr/bin/env python3
# Kostya Benchmark: collatz (Python)
# Longest Collatz (3n+1) sequence under 1,000,000
import time


def collatz_len(n):
    steps = 1
    x = n
    while x != 1:
        if x % 2 == 0:
            x //= 2
        else:
            x = 3 * x + 1
        steps += 1
    return steps


def benchmark():
    limit = 1000000
    max_len = 0
    max_start = 0
    for i in range(1, limit):
        clen = collatz_len(i)
        if clen > max_len:
            max_len = clen
            max_start = i
    return max_start


def main():
    t0 = time.perf_counter_ns()
    result = benchmark()
    t1 = time.perf_counter_ns()

    if result == 837799:
        print(f"collatz: PASS (start={result})")
    else:
        print(f"collatz: FAIL result={result}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()
